### PR TITLE
PEAT 1.1 (new formula w/ tests)

### PR DESCRIPTION
**Pull request by tseemann**
PEAT, a fast and highly accurate adapter-trimming algorithm, designed specifically for paired-end sequencing. PEAT requires no a priori adapter sequence and significantly outperforms other widely used adapter trimmers in both simulated and real life paired-end sequencing libraries. Moreover, in order to process single-end sequencing data, we also provide single-end trimmer in PEAT.

### CODE REVIEW

**Comment by tseemann** [comment]
Can anyone advise on this?
I'm guessing it's a BOOST / C++ / GCC issue?

```
/usr/local/Library/ENV/4.3/clang++   -Ofast -std=c++11 -stdlib=libc++ -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/bin/PEAT.dir/src/main.cpp.o  -o bin/PEAT  /usr/local/lib/libboost_date_time-mt.dylib /usr/local/lib/libboost_filesystem-mt.dylib /usr/local/lib/libboost_system-mt.dylib /usr/local/lib/libboost_serialization-mt.dylib /usr/local/lib/libboost_system-mt.dylib /usr/local/lib/libboost_filesystem-mt.dylib /usr/local/lib/libboost_regex-mt.dylib /usr/local/lib/libboost_thread-mt.dylib /usr/local/lib/libboost_iostreams-mt.dylib /usr/local/lib/libboost_program_options-mt.dylib /usr/lib/libz.dylib src/ThreadPool/libthread_pool.a /usr/local/lib/libboost_serialization-mt.dylib /usr/local/lib/libboost_regex-mt.dylib /usr/local/lib/libboost_thread-mt.dylib /usr/local/lib/libboost_iostreams-mt.dylib /usr/local/lib/libboost_program_options-mt.dylib /usr/lib/libz.dylib src/ThreadPool/libthread_pool.a 
Undefined symbols for architecture x86_64:
  "_GlobalPool", referenced from:
      paired_end::main(int, char**) in main.cpp.o
      single_end::main(int, char**) in main.cpp.o
      Job_distributer_pipeline<(ParallelTypes)1, std::__1::vector<Fastq<std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >, 
```

**Comment by tseemann** [comment]
```
 "boost::program_options::value_semantic_codecvt_helper<char>::parse(boost::any&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool) const", referenced from:
      vtable for boost::program_options::typed_value<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char> in main.cpp.o
      vtable for boost::program_options::typed_value<int, char> in main.cpp.o
ld: symbol(s) not found for architecture x86_64
```

File: peat.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  # tag "bioinformatics"
+  # doi "10.1186/1471-2105-16-S1-S2"
+
+  url "https://github.com/jhhung/PEAT/archive/v1.1.tar.gz"
+  sha256 "96d33f559e3602a8f01fef26ffb88f0e9f11d4d572d37e4db2e0702607aba32e"
+
+  depends_on "cmake" => :build
+  depends_on "boost"
+
+  fails_with :clang
```
Comment (xu-cheng) [inline peat.rb]: Try remove this line.

Comment (tseemann) [inline peat.rb]: Thanks @xu-cheng I will give that a try!

**Comment by xu-cheng** [comment]
Looks like it failed because its own vendored [libthread_pool.a](https://github.com/jhhung/PEAT/blob/master/src/ThreadPool/thread_pool.h#L81).

At least, I had no trouble to build PEAT 1.0b1 when you submit https://github.com/Homebrew/homebrew-science/pull/2156

**Comment by tseemann** [comment]
@xu-cheng Do you think this means we can never build it on Mac OSX ?

**Comment by xu-cheng** [comment]
No I think it's an upstream bug. Also the document clearly states that it supports OS X.

**Comment by tseemann** [comment]
@xu-cheng Would you be able to file an upstream bug with the PEAT authors, please?
I don't know much about C++/Boost!

**Comment by xu-cheng** [comment]
https://github.com/jhhung/PEAT/issues/21

**Comment by tseemann** [comment]
Thanks @xu-cheng , hopefully they will respond and fix it.

**Comment by xu-cheng** [comment]
Upstream has fixed the bug. Can you try to update the PR?

**Comment by tseemann** [comment]
```
IhandlerFactory.cpp:(.text._ZN7FDeviceI12SharedMemoryIN5boost3mpl3mapINS2_4pairIN4mpl_4int_ILi0EEENS1_9iostreams19filtering_streambufINS8_5inputEcSt11ch
ar_traitsIcESaIcENS8_7public_EEEEENS4_INS6_ILi1EEE15DeviceParameterEENS4_INS6_ILi3EEENS6_ILi2EEEEENS4_INS6_ILi4EEENS6_ILi8EEEEENS4_INS6_ILi5EEENS6_ILi16
EEEEENS4_INS6_ILi6EEESiEENS5_2naESV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_EEENS3_INS4_ISL_SQ_EESV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_EE
Li5EEC2ERSX_[_ZN7FDeviceI12SharedMemoryIN5boost3mpl3mapINS2_4pairIN4mpl_4int_ILi0EEENS1_9iostreams19filtering_streambufINS8_5inputEcSt11char_traitsIcESa
IcENS8_7public_EEEEENS4_INS6_ILi1EEE15DeviceParameterEENS4_INS6_ILi3EEENS6_ILi2EEEEENS4_INS6_ILi4EEENS6_ILi8EEEEENS4_INS6_ILi5EEENS6_ILi16EEEEENS4_INS6_
ILi6EEESiEENS5_2naESV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_EEENS3_INS4_ISL_SQ_EESV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_SV_EELi5EEC5ERSX_]+
0x28): undefined reference to `boost::iostreams::file_descriptor_source::file_descriptor_source(std::string const&, std::_Ios_Openmode)'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/PEAT] Error 1
```

**Comment by xu-cheng** [comment]
The CMakeLists.txt file in upstream is [incorrect](https://github.com/jhhung/PEAT/commit/0c18008966ca2340fc06108c000bea1b7c3af8bb#commitcomment-12356728).

In the mean time, you can do this:
```ruby
    inreplace "CMakeLists.txt", "libthread.a", "libthread_mac.a"
    inreplace "CMakeLists.txt", "IhandlerFactory.a", "IhandlerFactory_mac.a"
```

File: peat.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    cmake_args << "-DBOOST_ROOT=#{Formula["boost"].opt_prefix}"
+    system "cmake", ".", *cmake_args
+    system "make"
+    bin.install "bin/PEAT"
+    doc.install "README.md", "LICENSE"
+    (share/"peat").install Dir["test_file/*"]
+  end
+
+  test do
+    assert_match "trimming", shell_output("PEAT 2>&1", 1)
```
Comment (xu-cheng) [inline peat.rb]: `shell_output("PEAT 2>&1", 1)` should be `shell_output("#{bin}/PEAT 2>&1", 1)`

**Comment by xu-cheng** [comment]
PEAT 1.2 has release. It should have fixed all the problems

**Comment by tseemann** [comment]
I just tried to build PEAT 1.2.2 and still have the same problem:

```
CMakeFiles/bin/PEAT.dir/src/main.cpp.o: In function `paired_end::main(int, char**)':
main.cpp:(.text.startup+0xed6): undefined reference to `IhandlerFactory<Fastq, std::tuple<std::__cxx11::basic_string<cha
r, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocat
or<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<
char, std::char_traits<char>, std::allocator<char> > > >::get_ihandler_read(std::vector<std::__cxx11::basic_string<char,
 std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>,
 std::allocator<char> > > >, std::map<int, std::vector<Fastq<std::tuple<std::__cxx11::basic_string<char, std::char_trait
s<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::_
_cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_tr
aits<char>, std::allocator<char> > > >, std::allocator<Fastq<std::tuple<std::__cxx11::basic_string<char, std::char_trait
s<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::_
_cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_tr
aits<char>, std::allocator<char> > > > > >, std::less<int>, std::allocator<std::pair<int const, std::vector<Fastq<std::t
uple<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, s
td::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<c
har> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::allocator<Fastq<std::t
uple<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, s
td::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<c
har> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > > > >&)'
CMakeFiles/bin/PEAT.dir/src/main.cpp.o: In function `single_end::main(int, char**)':
```
